From 45a657b52573358a08613fcd9929de6f6b267cc6 Mon Sep 17 00:00:00 2001
From: Tiejun Chen <tiejun.chen@windriver.com>
Date: Thu, 13 Jan 2011 09:40:09 +0800
Subject: [PATCH] ppc-mc/direct-irq: to eoi spurious irqs when reboot

When reboot some devices may generate spurious irq so we have to
eoi that and skip this unnecessary interrupt process before kernel
can handle it.

Signed-off-by: Tiejun Chen <tiejun.chen@windriver.com>
---
 arch/powerpc/kernel/vbi/wrhv.c |    9 +++++++++
 1 files changed, 9 insertions(+), 0 deletions(-)

diff --git a/arch/powerpc/kernel/vbi/wrhv.c b/arch/powerpc/kernel/vbi/wrhv.c
index ba71064..b0da691 100644
--- a/arch/powerpc/kernel/vbi/wrhv.c
+++ b/arch/powerpc/kernel/vbi/wrhv.c
@@ -668,6 +668,15 @@ unsigned int wrhv_vioapic_get_irq(void)
 unsigned int wrhv_get_direct_irq(void)
 {
 	unsigned int irq = mfspr(SPRN_EPR);
+	/* When reboot some devices may generate spurious irq so here
+	 * eoi that and skip this unnecessary interrupt process.
+	 */
+	if (!irq_to_desc(irq)) {
+		printk(KERN_INFO "wrhv: spurious interrupt %d acknowledged.\n",
+			irq);
+		vbi_di_eoi();
+		irq = NO_IRQ;
+	}
 	return irq;
 }
 
-- 
1.6.5.2

